pandas/core/groupby/groupby.py

```diff
@@ -497,7 +497,8 @@ def _set_group_selection(self):
 
         if len(groupers):
             # GH12839 clear selected obj cache when group selection changes
-            self._group_selection = ax.difference(Index(groupers)).tolist()
+            self._group_selection = ax.difference(Index(groupers),
+                                                  sort=False).tolist()
```

Comment:
``Index.difference`` tries to sort its result by default and this means that sometimes the order of the columns was changed from the original DataFrame. I added a new ``sort`` parameter to ``Index.difference`` with a default of True to control this.